pageserver/src/layered_repository.rs

```diff
@@ -1743,8 +1758,33 @@ impl LayeredTimeline {
                 }
                 continue 'outer;
             }
-
-            // 2. Is it needed by a child branch?
+            // 2. It is newer than PiTR interval?
+            // We use modification time of layer file to estimate update time.
+            // This estimation is not quite precise but maintaining LSN->timestamp map seems to be overkill.
+            // It is not expected that users will need high precision here. And this estimation
+            // is conservative: modification time of file is always newer than actual time of version
+            // creation. So it is safe for users.
+            //
+            if let Ok(metadata) = fs::metadata(&l.filename()) {
+                let last_modified = metadata.modified()?;
```

Comment:
Also, what do you think about writing layer file creation time inside it? fs metadata can become out of sync with the real file creation time because of s3 upload/download for example. We won't loose anything, but this way we will keep potentially a lot of outdated files. 

Another idea, we can use some index file with timestamps for all layers so this way it is possible to avoid some extra  fs operations to make gc a bit cheaper.

Reply:
Modification time is greater or equal than creation time. As far as we are using it for cut-off, it is more conservative.
The semantic of PiTR interval is that user want to return back to the point which is not older in time than specified value. So using modification data (which correponds to last recent update) is safer.

Reply:
Sorry, I meant slightly different thing, I meant that both creation and modification times may not be preserved through full cycle of upload/download (I confirmed that with a local run of disabled on main test_tenant_relocation). e g file creation/modification time will be the time of the download, not the time of a file creation on the original pageserver it came from.

Also as far as we are using append only style creation vs modification does not make sense.

Reply:
First of all eviction procedure for pageserver data is not currently implemented (o I missed something).
So pageserver now contains all data and never request it from S3.
In theory it will be possible that we have independent GC (and PiTR intervals) for pageserver and S3.Butright now it is not the case.

But even if some tenant/timeline was just loaded to pageserver from S3, then current approach is safe: it will not remove data which should not be removed according to PiTR interval. Yes, it may not remove too old layer just because it was just uploaded from S3. And if this layer will be evicted and then once again uploaded to this pageserver, then in theory this layer may never be garbage collected. But IMHO it is very strange and unrealistic scenario.

Reply:
> First of all eviction procedure for pageserver data is not currently implemented (o I missed something).

Every checkpoint is uploaded to s3 now. It is not proper eviction but files are transferred and IMO it is enough to cause trouble (bloat)

> So pageserver now contains all data and never request it from S3.

New pageserver can request data from s3 e g during tenant migration.

> Yes, it may not remove too old layer just because it was just uploaded from S3. And if this layer will be evicted and then once again uploaded to this pageserver, then in theory this layer may never be garbage collected. But IMHO it is very strange and unrealistic scenario.

Thats exactly what I'm talking about. It is safe but we potentially may keep a lot of extra data causing bloat.
Eviction/on-demand download are planned to be implemented so this case is definitely realistic to me